### PR TITLE
Feature/feature crud medals

### PR DESCRIPTION
**Pull request by oricat8**

### CODE REVIEW

**Comment by santidrj** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 7
           

Complexity increasing per file
==============================
- usermanager/src/main/java/org/pesmypetcare/usermanager/datacontainers/user/UserMedalData.java  4
- usermanager/src/main/java/org/pesmypetcare/usermanager/clients/user/UserMedalManagerClient.java  1
- usermanager/src/test/java/org/pesmypetcare/usermanager/clients/user/UserMedalManagerClientTest.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/Grupo13-PES-Mascotas/PES_My-Pet-Care-Libraries/pullRequest?prid=5633147&bid=18448897)

File: usermanager/src/main/java/org/pesmypetcare/usermanager/datacontainers/user/UserMedalData.java [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+                && Arrays.equals(medalIcon, that.medalIcon)
+                && Objects.equals(progress, that.progress)
+                && Objects.equals(currentLevel, that.currentLevel)
+                && Objects.equals(completedLevelsDate, that.completedLevelsDate);
+    }
+
+    @Override
+    public int hashCode() {
+        int result = Objects.hash(name, levels, description, progress, currentLevel, completedLevelsDate);
+        result = 31 * result + Arrays.hashCode(medalIcon);
```
Comment (santidrj) [inline usermanager/src/main/java/org/pesmypetcare/usermanager/datacontainers/user/UserMedalData.java:113]: Codacy Issue found: ['31' is a magic number.](https://app.codacy.com/gh/Grupo13-PES-Mascotas/PES_My-Pet-Care-Libraries/pullRequest?prid=5633147)

File: usermanager/src/test/java/org/pesmypetcare/usermanager/clients/user/UserMedalManagerClientTest.java [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+        headers = new HashMap<>();
+        headers.put(token, token);
+    }
+
+    @Test
+    public void getMedal() throws MyPetCareException {
+        HttpParameter[] params = new HttpParameter[1];
+        params[0] = new HttpParameter(USERMEDAL_PARAMETER, medalName);
+        given(httpClient
+                .get(eq(USERMEDAL_PATH + ownerNameEncoded + "/" + medalNameEncoded),
```
Comment (santidrj) [inline usermanager/src/test/java/org/pesmypetcare/usermanager/clients/user/UserMedalManagerClientTest.java:85]: Codacy Issue found: [The String "/" appears 5 times in the file.](https://app.codacy.com/gh/Grupo13-PES-Mascotas/PES_My-Pet-Care-Libraries/pullRequest?prid=5633147)

File: usermanager/src/test/java/org/pesmypetcare/usermanager/clients/user/UserMedalManagerClientTest.java [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+        gson = new Gson();
+        owner = "Rosa Melano";
+        medalName = "Walker";
+        token = "token";
+        field = "progress";
+        fieldNameEncoded = HttpParameter.encode(field);
+        ownerNameEncoded = HttpParameter.encode(owner);
+        medal = new UserMedalData("Walker",
+                new ArrayList<>(Arrays.asList(5., 10., 25., 50., 100.)),
+                "You have to walk a lot!", new byte[] {10, 55, 67, 89, 103, 116},
```
Comment (santidrj) [inline usermanager/src/test/java/org/pesmypetcare/usermanager/clients/user/UserMedalManagerClientTest.java:72]: Codacy Issue found: ['10' is a magic number.](https://app.codacy.com/gh/Grupo13-PES-Mascotas/PES_My-Pet-Care-Libraries/pullRequest?prid=5633147)

File: usermanager/src/main/java/org/pesmypetcare/usermanager/datacontainers/user/UserMedalData.java [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+    public List<Double> getLevels() {
+        return levels;
+    }
+
+    public String getDescription() {
+        return description;
+    }
+
+    public byte[] getMedalIcon() {
+        return medalIcon;
```
Comment (santidrj) [inline usermanager/src/main/java/org/pesmypetcare/usermanager/datacontainers/user/UserMedalData.java:63]: Codacy Issue found: [Returning 'medalIcon' may expose an internal array.](https://app.codacy.com/gh/Grupo13-PES-Mascotas/PES_My-Pet-Care-Libraries/pullRequest?prid=5633147)

File: usermanager/src/test/java/org/pesmypetcare/usermanager/clients/user/UserMedalManagerClientTest.java [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+    private HttpResponse httpResponse;
+
+    @InjectMocks
+    private UserMedalManagerClient client = new UserMedalManagerClient();
+
+    @Before
+    public void setUp() {
+        gson = new Gson();
+        owner = "Rosa Melano";
+        medalName = "Walker";
```
Comment (santidrj) [inline usermanager/src/test/java/org/pesmypetcare/usermanager/clients/user/UserMedalManagerClientTest.java:65]: Codacy Issue found: [The String "Walker" appears 2 times in the file.](https://app.codacy.com/gh/Grupo13-PES-Mascotas/PES_My-Pet-Care-Libraries/pullRequest?prid=5633147)

File: usermanager/src/main/java/org/pesmypetcare/usermanager/datacontainers/user/UserMedalData.java [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+     * @param name UserMedalData's name
+     * @param levels Goals of the UserMedalData's levels
+     * @param description UserMedalData's description
+     * @param medalIcon UserMedalData's icon
+     * @param progress UserMedalData's progress
+     * @param currentLevel UserMedalData's current level of the medal
+     * @param completedLevelsDate UserMedalData's array with the completed levels dates.
+     */
+    public UserMedalData(String name, List<Double> levels, String description,
+                         byte[] medalIcon, Double progress, Double currentLevel,
```
Comment (santidrj) [inline usermanager/src/main/java/org/pesmypetcare/usermanager/datacontainers/user/UserMedalData.java:39]: Codacy Issue found: [The user-supplied array 'medalIcon' is stored directly.](https://app.codacy.com/gh/Grupo13-PES-Mascotas/PES_My-Pet-Care-Libraries/pullRequest?prid=5633147)

File: usermanager/src/test/java/org/pesmypetcare/usermanager/clients/user/UserMedalManagerClientTest.java [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+        token = "token";
+        field = "progress";
+        fieldNameEncoded = HttpParameter.encode(field);
+        ownerNameEncoded = HttpParameter.encode(owner);
+        medal = new UserMedalData("Walker",
+                new ArrayList<>(Arrays.asList(5., 10., 25., 50., 100.)),
```
Comment (santidrj) [inline usermanager/src/test/java/org/pesmypetcare/usermanager/clients/user/UserMedalManagerClientTest.java:71]: Codacy Issue found: ['5.' is a magic number.](https://app.codacy.com/gh/Grupo13-PES-Mascotas/PES_My-Pet-Care-Libraries/pullRequest?prid=5633147)

File: usermanager/src/main/java/org/pesmypetcare/usermanager/datacontainers/user/UserMedalData.java (hunk 6 above)
Comment (marclll) [reply usermanager/src/main/java/org/pesmypetcare/usermanager/datacontainers/user/UserMedalData.java:39]: Has de corregir aquest error

File: usermanager/src/main/java/org/pesmypetcare/usermanager/datacontainers/user/UserMedalData.java (hunk 4 above)
Comment (marclll) [reply usermanager/src/main/java/org/pesmypetcare/usermanager/datacontainers/user/UserMedalData.java:63]: Aquest l'has de corregir també

Comment (marclll) [reply usermanager/src/main/java/org/pesmypetcare/usermanager/datacontainers/user/UserMedalData.java:63]: Has de retornar una copia

File: usermanager/src/main/java/org/pesmypetcare/usermanager/datacontainers/user/UserMedalData.java (hunk 6 above)
Comment (marclll) [reply usermanager/src/main/java/org/pesmypetcare/usermanager/datacontainers/user/UserMedalData.java:39]: Has de guardar una copia, ja que java treballa amb referencies

File: usermanager/src/main/java/org/pesmypetcare/usermanager/datacontainers/user/UserMedalData.java (hunk 1 above)
Comment (marclll) [reply usermanager/src/main/java/org/pesmypetcare/usermanager/datacontainers/user/UserMedalData.java:113]: Perque 31?

File: usermanager/src/test/java/org/pesmypetcare/usermanager/clients/user/UserMedalManagerClientTest.java [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+    @Before
+    public void setUp() {
+        gson = new Gson();
+        owner = "Rosa Melano";
+        medalName = "Walker";
+        token = "token";
+        field = "progress";
+        fieldNameEncoded = HttpParameter.encode(field);
+        ownerNameEncoded = HttpParameter.encode(owner);
+        medal = new UserMedalData("Walker",
```
Comment (marclll) [inline usermanager/src/test/java/org/pesmypetcare/usermanager/clients/user/UserMedalManagerClientTest.java:70]: Si guardes el nom en una variable no el fiques després hardcoded

File: usermanager/src/test/java/org/pesmypetcare/usermanager/clients/user/UserMedalManagerClientTest.java [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+    public void setUp() {
+        gson = new Gson();
+        owner = "Rosa Melano";
+        medalName = "Walker";
+        token = "token";
+        field = "progress";
+        fieldNameEncoded = HttpParameter.encode(field);
+        ownerNameEncoded = HttpParameter.encode(owner);
+        medal = new UserMedalData(medalName,
+                new ArrayList<>(Arrays.asList(5., 10., 25., 50., 100.)),
```
Comment (santidrj) [inline usermanager/src/test/java/org/pesmypetcare/usermanager/clients/user/UserMedalManagerClientTest.java:71]: Codacy Issue found: ['5.' is a magic number.](https://app.codacy.com/gh/Grupo13-PES-Mascotas/PES_My-Pet-Care-Libraries/pullRequest?prid=5633147)

File: usermanager/src/main/java/org/pesmypetcare/usermanager/datacontainers/user/UserMedalData.java [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+                         List<String> completedLevelsDate) {
+        this.name = name;
+        this.levels = levels;
+        this.description = description;
+        if (medalIcon != null) {
+            this.medalIcon = Arrays.copyOf(medalIcon, medalIcon.length);
+        } else {
+            this.medalIcon = null;
```
Comment (santidrj) [inline usermanager/src/main/java/org/pesmypetcare/usermanager/datacontainers/user/UserMedalData.java:48]: Codacy Issue found: [Assigning an Object to null is a code smell.  Consider refactoring.](https://app.codacy.com/gh/Grupo13-PES-Mascotas/PES_My-Pet-Care-Libraries/pullRequest?prid=5633147)

**Comment by santidrj** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 3
           

Complexity increasing per file
==============================
- usermanager/src/main/java/org/pesmypetcare/usermanager/datacontainers/user/UserMedalData.java  4
- usermanager/src/main/java/org/pesmypetcare/usermanager/clients/user/UserMedalManagerClient.java  1
- usermanager/src/test/java/org/pesmypetcare/usermanager/clients/user/UserMedalManagerClientTest.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/Grupo13-PES-Mascotas/PES_My-Pet-Care-Libraries/pullRequest?prid=5633147&bid=18448897)

File: usermanager/src/test/java/org/pesmypetcare/usermanager/clients/user/UserMedalManagerClientTest.java [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+                "You have to walk a lot!", new byte[] {10, 55, 67, 89, 103, 116},
```
Comment (santidrj) [inline usermanager/src/test/java/org/pesmypetcare/usermanager/clients/user/UserMedalManagerClientTest.java:72]: Codacy Issue found: ['10' is a magic number.](https://app.codacy.com/gh/Grupo13-PES-Mascotas/PES_My-Pet-Care-Libraries/pullRequest?prid=5633147)

File: usermanager/src/test/java/org/pesmypetcare/usermanager/clients/user/UserMedalManagerClientTest.java [hunk 12]
```diff
@@ -0,0 +1,131 @@
...
+        headers = new HashMap<>();
+        headers.put(token, token);
+    }
+
+    @Test
+    public void getMedal() throws MyPetCareException {
+        HttpParameter[] params = new HttpParameter[1];
+        params[0] = new HttpParameter(USERMEDAL_PARAMETER, medalName);
+        given(httpClient
+                .get(eq(USERMEDAL_PATH + ownerNameEncoded + "/" + medalNameEncoded),
```
Comment (santidrj) [inline usermanager/src/test/java/org/pesmypetcare/usermanager/clients/user/UserMedalManagerClientTest.java:85]: Codacy Issue found: [The String "/" appears 5 times in the file.](https://app.codacy.com/gh/Grupo13-PES-Mascotas/PES_My-Pet-Care-Libraries/pullRequest?prid=5633147)

File: usermanager/src/test/java/org/pesmypetcare/usermanager/clients/user/UserMedalManagerClientTest.java (hunk 9 above)
Comment (santidrj) [inline usermanager/src/test/java/org/pesmypetcare/usermanager/clients/user/UserMedalManagerClientTest.java:71]: Codacy Issue found: ['5.' is a magic number.](https://app.codacy.com/gh/Grupo13-PES-Mascotas/PES_My-Pet-Care-Libraries/pullRequest?prid=5633147)

**Comment by santidrj** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- usermanager/src/test/java/org/pesmypetcare/usermanager/clients/user/UserMedalManagerClientTest.java  1
- usermanager/src/main/java/org/pesmypetcare/usermanager/datacontainers/user/UserMedalData.java  4
- usermanager/src/main/java/org/pesmypetcare/usermanager/clients/user/UserMedalManagerClient.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/Grupo13-PES-Mascotas/PES_My-Pet-Care-Libraries/pullRequest?prid=5633147&bid=18448897)

**Comment by santidrj** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- usermanager/src/test/java/org/pesmypetcare/usermanager/clients/user/UserMedalManagerClientTest.java  1
- usermanager/src/main/java/org/pesmypetcare/usermanager/clients/user/UserMedalManagerClient.java  1
- usermanager/src/main/java/org/pesmypetcare/usermanager/datacontainers/user/UserMedalData.java  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/Grupo13-PES-Mascotas/PES_My-Pet-Care-Libraries/pullRequest?prid=5633147&bid=18448897)

**Pull request merged**